### PR TITLE
fix: check for cilium system application in initial application controller

### LINKED ISSUE #12820: App installation at cluster setup can block CNI deployment
### What happened?

When creating a user cluster through the dashboard, step 5 can be used to configure Applications that will be deployed upon cluster creation. An obvious candidate are the apps from the default app catalog (EE feature).

Unfortunately, this feature is problematic. App installation waits for the Helm chart to succeed deployment, which means it waits for Pods to start up. But Pods will never start up when no CNI is deployed, because nodes are not fully initialised.

If Cilium 1.13+ is used as CNI, it is also deployed via an "System" Application. But since System Applications  do not seem to have priority over normal Applications, you can run into the situation that user-cluster-controller-manager is stuck trying to deploy a "random" Application (e.g. cert-manager), and never gets to deploy Cilium. The cluster is then stuck as CNI never starts, so regular Pods can never start.

Screenshot 2023-11-07 at 08 34 57

The `cert-manager` ApplicationInstallation has the following status:

```yaml
  status:
    applicationVersion:
      template:
        source:
          helm:
            chartName: cert-manager
            chartVersion: v1.12.3
            url: https://charts.jetstack.io
      version: v1.12.3
    conditions:
      ManifestsRetrieved:
        lastHeartbeatTime: "2023-11-07T07:31:24Z"
        message: application's source successfully downloaded
        observedGeneration: 1
        reason: DownloadSourceSuccessful
        status: "True"
      Ready:
        lastHeartbeatTime: "2023-11-07T07:31:24Z"
        message: application is installing or upgrading
        observedGeneration: 1
        reason: InstallationInProgress
        status: Unknown
    method: helm
```

### Expected behavior

Cilium CNI "System Application" is installed with priority over normal Applications.

### How to reproduce the issue?

1. Create user cluster, during the wizard select "cert-manager" as Application to deploy.
2. Observe "cert-manager" Application stuck in deployment and "Cilium" Application never actually deploy.

### How is your environment configured?

- KKP version: `main`
- Shared or separate master/seed clusters?: shared

### Provide your KKP manifest here (if applicable)

<details>

```yaml
# paste manifest here
```

</details>

### What cloud provider are you running on?

N/A

### What operating system are you running in your user cluster?

N/A

### PR DESCRIPTION
**Pull request by cnvergence**
**What this PR does / why we need it**:

**Which issue(s) this PR fixes**:

Fixes https://github.com/kubermatic/kubermatic/issues/12820

**What type of PR is this?**

/kind bug

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change? Then add your Release Note here**:

```release-note
Fixing the issue with blocked cluster provisioning, when selected initial applications that conflicted with Cilium system application and user-cluster-controller-manager was stuck.
```

**Documentation**:

```documentation
NONE
```

### CODE REVIEW

**Comment by cnvergence** [comment]
reproduced the issue on the seperate run:
- https://github.com/kubermatic/kubermatic/pull/13007
- https://public-prow.loodse.com/view/gs/prow-dev-public-data/pr-logs/pull/kubermatic_kubermatic/13007/pre-kubermatic-cilium-e2e/1752332248641703936

File: pkg/controller/seed-controller-manager/initial-application-installation-controller/controller.go [hunk 1]
```diff
@@ -224,3 +240,12 @@ func (r *Reconciler) removeAnnotation(ctx context.Context, cluster *kubermaticv1
 	delete(cluster.Annotations, kubermaticv1.InitialApplicationInstallationsRequestAnnotation)
 	return r.Patch(ctx, cluster, ctrlruntimeclient.MergeFrom(oldCluster))
 }
+
+func getCiliumApplicationInstallation(userClusterClient ctrlruntimeclient.Client) (*appskubermaticv1.ApplicationInstallation, error) {
+	app := &appskubermaticv1.ApplicationInstallation{}
+	if err := userClusterClient.Get(context.Background(), types.NamespacedName{Namespace: metav1.NamespaceSystem, Name: kubermaticv1.CNIPluginTypeCilium.String()}, app); err != nil {
```
Comment (embik) [inline pkg/controller/seed-controller-manager/initial-application-installation-controller/controller.go]: Can we please make sure that the context here is not `context.Background()` but `ctx` from the reconcile loop?

File: pkg/controller/seed-controller-manager/initial-application-installation-controller/controller.go [hunk 2]
```diff
@@ -224,3 +240,12 @@ func (r *Reconciler) removeAnnotation(ctx context.Context, cluster *kubermaticv1
 	delete(cluster.Annotations, kubermaticv1.InitialApplicationInstallationsRequestAnnotation)
 	return r.Patch(ctx, cluster, ctrlruntimeclient.MergeFrom(oldCluster))
 }
+
+func getCiliumApplicationInstallation(userClusterClient ctrlruntimeclient.Client) (*appskubermaticv1.ApplicationInstallation, error) {
```
Comment (embik) [inline pkg/controller/seed-controller-manager/initial-application-installation-controller/controller.go]: I'd love for this to be more generic. I know that _at the moment_ it's only Cilium, but I could see us move more CNIs to the app infrastructure in the future.

I think this should be:

```suggestion
func getCNIApplicationInstallation(ctx context.Context, userClusterClient ctrlruntimeclient.Client, cniType kubermaticv1.CNIPluginType) (*appskubermaticv1.ApplicationInstallation, error) {
```

There could be a switch statement in here which only has a case for `kubermaticv1.CNIPluginTypeCilium` but could easily be extended later on.

**Review by embik — CHANGES_REQUESTED** [review]
Great PR, minor suggested changes for future extensibility.

File: pkg/controller/seed-controller-manager/initial-application-installation-controller/controller.go (hunk 2 above)
Comment (cnvergence) [reply pkg/controller/seed-controller-manager/initial-application-installation-controller/controller.go]: good point, let's do it beforehand

File: pkg/controller/seed-controller-manager/initial-application-installation-controller/controller.go [hunk 3]
```diff
@@ -224,3 +240,16 @@ func (r *Reconciler) removeAnnotation(ctx context.Context, cluster *kubermaticv1
...
+
+func getCNIApplicationInstallation(ctx context.Context, userClusterClient ctrlruntimeclient.Client, cniType kubermaticv1.CNIPluginType) (*appskubermaticv1.ApplicationInstallation, error) {
+	app := &appskubermaticv1.ApplicationInstallation{}
+	switch cniType {
+	case kubermaticv1.CNIPluginTypeCilium:
+		name := kubermaticv1.CNIPluginTypeCilium.String()
+		if err := userClusterClient.Get(ctx, types.NamespacedName{Namespace: metav1.NamespaceSystem, Name: name}, app); err != nil {
+			return nil, fmt.Errorf("failed to get Cilium ApplicationInstallation in user cluster: %w", err)
+		}
+	}
```
Comment (embik) [inline pkg/controller/seed-controller-manager/initial-application-installation-controller/controller.go]: Sorry, one more addition here: Can you please add a `default` case that returns an "unsupported" error? Else we return the `nil` pointer `app` variable in the end, don't think we should do that.

**Review by embik — CHANGES_REQUESTED** [review]

File: pkg/controller/seed-controller-manager/initial-application-installation-controller/controller.go (hunk 3 above)
Comment (cnvergence) [reply pkg/controller/seed-controller-manager/initial-application-installation-controller/controller.go]: Hmm, sure we can do that, we are checking earlier if cni.IsManagedByAppInfra.
Also we can just return this unsupported error after switch statement

Comment (embik) [reply pkg/controller/seed-controller-manager/initial-application-installation-controller/controller.go]: I agree that it shouldn't happen, but IMO it's a good idea to do it anyway to exhaust all possible inputs. I don't have strong opinions whether it's in a `default` block or after the switch statement.

**Review by embik — APPROVED** [review]
/approve

thank you!

**Pull request merged**